src/main/java/com/google/sps/data/DataStoreManager.java

```diff
@@ -0,0 +1,338 @@
+    String gameCreator = (String) gameEntity.getProperty("gameCreator");
+    int numStages = (int)gameEntity.getProperty("numStages");
+    ArrayList<String> stages = (ArrayList<String>)gameEntity.getProperty("stages");
+    int numTimesPlayed = (int)gameEntity.getProperty("numTimesPlayed");
+    int numTimesFinished = (int)gameEntity.getProperty("numTimesFinished");
+    int numStarVotes = (int)gameEntity.getProperty("numStarVotes");
+    int totalStars = (int)gameEntity.getProperty("totalStars");
+    int numDifficultyVotes = (int)gameEntity.getProperty("numDifficultyVotes");
+    int totalDifficulty = (int)gameEntity.getProperty("totalDifficulty");
+  
+    Game.Builder game = new Game.Builder(gameID, gameName).setGameCreator(gameCreator).setNumStages(numStages);
+    game.setGameDescription(gameDescription).setNumTimesPlayed(numTimesPlayed).setNumTimesFinished(numTimesFinished);
+    game.setNumStarVotes(numStarVotes).setTotalStars(totalStars).setNumDifficultyVotes(numDifficultyVotes).setTotalDifficulty(totalDifficulty);       
+    return game.build();
+  }
+
+  /**
+    * Stores the static data of a single stage in datastore as an entity
+    * @param game an Stage variable representing a single instance of a stage.
+  */
+  public void storeStage(Stage stage) {
+    Entity stageEntity = new Entity("Stage", stage.getStageID());
+    stageEntity.setProperty("stageNumber", stage.getStageNumber());
+    stageEntity.setProperty("key", stage.getKey());
+    stageEntity.setProperty("startingHint", stage.getStartingHint());
+    stageEntity.setProperty("latitude", stage.getStartingLocation().getLatitude());
+    stageEntity.setProperty("longitude", stage.getStartingLocation().getLongitude());
+    for (Hint hint : stage.getHints()) {
+      storeHint(hint);
+    }
+    ArrayList<String> hints;
+    for (Hint hint : stage.getHints()) {
+      hints.add(hint.getHintID());
+    }
+    stageEntity.setProperty("hints", hints);
+    datastore.put(stageEntity);
+  }
+  
+  /**
+    * Updates an entity of a single stage data in datastore
+    * @param stage a Stage variable representing a single instance of a stage.
+  */
+  public void updateStage(Stage stage) {
+    Key stageEntityKey = KeyFactory.createKey("Stage", stage.getStageID());
+    Entity stageEntity;
+    try {
+      stageEntity = datastore.get(stageEntityKey);
+    }
+    catch(Exception e) {
+      return;
+    }
+      
+    stageEntity.setProperty("stageNumber", stage.getStageNumber());
+    stageEntity.setProperty("key", stage.getKey());
+    stageEntity.setProperty("startingHint", stage.getStartingHint());
+    stageEntity.setProperty("latitude", stage.getStartingLocation().getLatitude());
+    stageEntity.setProperty("longitude", stage.getStartingLocation().getLongitude());
+    for (Hint hint : stage.getHints()) {
+      updateHint(hint);
+    }
+    ArrayList<String> hints;
+    for (Hint hint : stage.getHints()) {
```

Comment:
Again, you can combine this loop with the one above it.

Reply:
will do